### PR TITLE
OTC Desk Contract

### PR DESCRIPTION
**Pull request by katelynsills**
Inspired by the signature-based pricing mentioned in https://medium.com/dragonfly-research/unbundling-uniswap-the-future-of-on-chain-market-making-1c7d6948d570

The creator of the contract instance gets a creatorFacet that allows them to add inventory, remove inventory, make quotes and create specific invitations to exercise those quotes. A user who receives an invitation can look at the invitationDetails for the quoted trade, and using the keywords of their choice, accept the quote and offer the required digital assets to complete the exchange.

This version uses a covered call instance to perform the escrowing and trade.

### CODE REVIEW

File: packages/zoe/src/contracts/otcDesk.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+import { E } from '@agoric/eventual-send';
+import { assert, details } from '@agoric/assert';
+import { trade } from '../contractSupport';
+
+import '../../exported';
+
+/**
+ * The creator gets a creatorFacet that allows them to add inventory, make quotes and
+ * create specific invitations to exercise those quotes.
```
Comment (Chris-Hibbert) [inline packages/zoe/src/contracts/otcDesk.js]: Please add more details on the usage of this contract. What shape proposals are accepted? What methods are in the creatorFacet? What do the invitations that the creator can hand out allow? What can recipients of those invitations rely on? 
`makeQuote` seems to have some interesting functionality. How is it used? 
Can the recipient of a `removeInventory` inventory verify that the inventory is reserved for their use?

File: packages/zoe/test/unitTests/contracts/test-otcDesk.js [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+  const { moolaPayment, simoleanPayment, bucksPayment } = origPayments;
+  const moolaPurse = moolaIssuer.makeEmptyPurse();
+  const simoleanPurse = simoleanIssuer.makeEmptyPurse();
+  const bucksPurse = bucksIssuer.makeEmptyPurse();
+
+  moolaPurse.deposit(moolaPayment);
+  simoleanPurse.deposit(simoleanPayment);
+  bucksPurse.deposit(bucksPayment);
+  return harden({
+    offerGood: async untrustedInvitation => {
```
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/contracts/test-otcDesk.js]: This name seems backwards. Bob's method should reflect that he's receiving an invitation

File: packages/zoe/test/unitTests/contracts/test-otcDesk.js [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+      const invitationIssuer = await E(zoe).getInvitationIssuer();
+
+      // Bob is able to use the trusted invitationIssuer from Zoe to
+      // transform an untrusted invitation that Alice also has access to, to
+      // an
```
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/contracts/test-otcDesk.js]: unfinished

File: packages/zoe/src/contracts/otcDesk.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    deadline,
+  ) => {
+    let quoteActive = true;
+    E(timerAuthority).setWakeup(
+      deadline,
+      harden({ wake: () => (quoteActive = false) }),
+    );
+    const acceptQuote = seat => {
+      assert(quoteActive, details`The quote was no longer active`);
+      const { want: userWantActual, give: userGiveActual } = seat.getProposal();
```
Comment (Chris-Hibbert) [inline packages/zoe/src/contracts/otcDesk.js]: This doesn't seem to match the model in the linked article. I thought the main point of the signed price quotes was that they were intended to be commitments to trade at the stated price. This contract will trade at that price if it has remaining funds, but there's no commitment to the counterparty that the funds are available.

File: packages/zoe/test/unitTests/contracts/test-otcDesk.js [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+  const invitation2 = await alice.makeQuoteForBob(
+    { Simolean: simoleans(4) },
+    { Moola: moola(3) },
+    timer,
+    1,
+  );
+  timer.tick();
+
+  // Bob tries to offer but the quote is expired.
+  await bob.offerExpired(invitation2);
```
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/contracts/test-otcDesk.js:486]: This can be a separate test case.

File: packages/zoe/test/unitTests/contracts/test-otcDesk.js [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+
+  const invitation3 = await alice.makeQuoteForBob(
+    { Buck: bucks(500), Moola: moola(35) },
+    { Simolean: simoleans(15) },
+    timer,
+    100,
+  );
+
+  // Bob tries to offer but he wants more than what was quoted.
+  await bob.offerWantTooMuch(invitation3);
```
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/contracts/test-otcDesk.js:534]: This can be a separate test case.

File: packages/zoe/test/unitTests/contracts/test-otcDesk.js [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+});
```
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/contracts/test-otcDesk.js:535]: As it seems to be part fo the design, I'd like to see tests where
* Alice makes a quote for Bob that isn't yet funded. She might or might not fund it before Bob exercises. 
* Alice makes a quote that is currently funded, but removes the funding before Bob receives it.

File: packages/zoe/test/swingsetTests/zoe/test-zoe.js [hunk 8]
```diff
@@ -285,6 +286,26 @@ test.serial('zoe - sellTickets - valid inputs', async t => {
...
+  '=> alice, bob, carol and dave are set up',
+  'Inventory added',
+  '@@ schedule task for:1, currently: 0 @@',
+  'Trade successful',
+  '{"brand":{},"value":3}',
+  '{"brand":{},"value":0}',
+  'Inventory removed',
+  '{"brand":{},"value":2}',
+];
+test.serial('zoe - otcDesk - valid inputs', async t => {
```
Comment (Chris-Hibbert) [inline packages/zoe/test/swingsetTests/zoe/test-zoe.js]: Why does it need to be serial? I suspect we should document these annotations when we add them.

File: packages/zoe/test/swingsetTests/zoe/test-zoe.js [hunk 9]
```diff
@@ -285,6 +286,26 @@ test.serial('zoe - sellTickets - valid inputs', async t => {
...
+  t.plan(1);
```
Comment (Chris-Hibbert) [inline packages/zoe/test/swingsetTests/zoe/test-zoe.js]: I thought we didn't need `t.plan()` with AVA. Why is this needed?

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]
This contract does what it claims, but it doesn't seem like an implementation of the OTC concept in the linked article. That article talks about signing legally binding OTC agreements, while this contract makes offers that might or might not be backed. The acceptor will get the offered deal when the contract has funding, but if not, they get their money back.

File: packages/zoe/test/swingsetTests/zoe/test-zoe.js (hunk 9 above)
Comment (katelynsills) [reply packages/zoe/test/swingsetTests/zoe/test-zoe.js]: I was cargo-culting (actually, more like copy-and-pasting). It looks like it got partially removed from the other tests in this file but not all. It is not necessary.

File: packages/zoe/test/swingsetTests/zoe/test-zoe.js (hunk 8 above)
Comment (katelynsills) [reply packages/zoe/test/swingsetTests/zoe/test-zoe.js]: This was done by Brian in this commit. My test does not do anything special here: https://github.com/Agoric/agoric-sdk/commit/c5364b8be273e3e307d8f5d4ef947047a725c702

File: packages/zoe/test/unitTests/contracts/test-otcDesk.js (hunk 2 above)
Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-otcDesk.js]: Well, it's the name of what action Bob is taking. In this case he is offering without triggering an error (offer - good). I can see why the name isn't great because goods are also a thing.

File: packages/zoe/src/contracts/otcDesk.js (hunk 4 above)
Comment (katelynsills) [reply packages/zoe/src/contracts/otcDesk.js]: Really good point. Fixed

File: packages/zoe/test/unitTests/contracts/test-otcDesk.js (hunk 6 above)
Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-otcDesk.js:534]: Done

File: packages/zoe/test/unitTests/contracts/test-otcDesk.js (hunk 7 above)
Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-otcDesk.js:535]: > Alice makes a quote that is currently funded, but removes the funding before Bob receives it.

This will be impossible in the new design, since the funding is escrowed in the covered call.

Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-otcDesk.js:535]: > Alice makes a quote for Bob that isn't yet funded. She might or might not fund it before Bob exercises.

This is also impossible in the new design.

File: packages/zoe/src/contracts/otcDesk.js [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+ * deadline. `makeQuote` returns a covered call option that can be
+ * given away for free or sold. Importantly, if the recipient chooses
+ * to exercise the option, the trade is guaranteed to succeed if their
+ * proposal matches the quote.
+ *
+ * @type {ContractStartFn}
+ */
+const start = zcf => {
+  const { zcfSeat: marketMakerSeat } = zcf.makeEmptySeatKit();
+  const { coveredCallInstallation } = zcf.getTerms();
```
Comment (dtribble) [inline packages/zoe/src/contracts/otcDesk.js:55]: Move up to first line in function since it's approximately parameters of the contract.

File: packages/zoe/src/contracts/otcDesk.js [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  const zoe = zcf.getZoeService();
+
+  const makeQuote = async (price, assets, timeAuthority, deadline) => {
```
Comment (dtribble) [inline packages/zoe/src/contracts/otcDesk.js:69]: Will it make the VSCode dev experience better if you declare the parameter types early?

File: packages/zoe/src/contracts/otcDesk.js [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+      payments,
+    );
+
+    E(sellerUserSeat)
+      .getPayouts()
+      .then(async payoutPayments => {
+        const amounts = await E(sellerUserSeat).getCurrentAllocation();
+        await depositToSeat(zcf, marketMakerSeat, amounts, payoutPayments);
+      });
+    const option = E(sellerUserSeat).getOfferResult();
```
Comment (dtribble) [inline packages/zoe/src/contracts/otcDesk.js]: Just return this directly?

Comment (dtribble) [reply packages/zoe/src/contracts/otcDesk.js]: Hmm I guess there's some value in naming it, but at least delete the intervening white space. But otherwise:
```
// return the option
return E(sellerUserSeat).getOfferResult();
```

File: packages/zoe/src/contracts/otcDesk.js [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+  };
+
+  const creatorFacet = {
+    // The inventory can be added in bulk before any quotes are made
+    // or can be added immediately before a quote.
+    makeAddInventoryInvitation: async (issuerKeywordRecord = {}) => {
+      await saveAllIssuers(zcf, issuerKeywordRecord);
+      return zcf.makeInvitation(addInventory, 'addInventory');
+    },
+    makeRemoveInventoryInvitation: () =>
```
Comment (dtribble) [inline packages/zoe/src/contracts/otcDesk.js]: I would add braces so it's structure is the same asa the other invite method

**Review by dtribble — APPROVED** [review]

File: packages/zoe/src/contracts/otcDesk.js (hunk 1 above)
Comment (katelynsills) [reply packages/zoe/src/contracts/otcDesk.js]: Done

File: packages/zoe/src/contracts/otcDesk.js (hunk 10 above)
Comment (katelynsills) [reply packages/zoe/src/contracts/otcDesk.js:55]: Done

File: packages/zoe/src/contracts/otcDesk.js (hunk 11 above)
Comment (katelynsills) [reply packages/zoe/src/contracts/otcDesk.js:69]: I think so, I will add those.

File: packages/zoe/src/contracts/otcDesk.js [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+ * `E(creatorFacet).makeQuote(price, assets, timeAuthority,
+ * deadline)`. `price` and `assets` are amountKeywordRecords that will
+ * be used in the coveredCall. The assets are the underlyingAssets in
+ * the call option, and the price is used as the strikePrice. The
+ * timeAuthority should be a timer, and the deadline can be any time
+ * understood by the timer. The quote will be cancelled after the
+ * deadline. `makeQuote` returns a covered call option that can be
+ * given away for free or sold. Importantly, if the recipient chooses
+ * to exercise the option, the trade is guaranteed to succeed if their
+ * proposal matches the quote.
```
Comment (Chris-Hibbert) [inline packages/zoe/src/contracts/otcDesk.js]: And if they are thinking of buying it, they can verify that the goods being offered are already escrowed,

**Review by Chris-Hibbert — APPROVED** [review]
sorry, neglected to approve.  This looks good to me. Nice work.

**Comment by katelynsills** [comment]
Merging as good enough. Any more changes will have to happen after tomorrow morning. Thanks!
